File: internal/store/sshkeys.go
package store

import (
	"database/sql"
	"errors"

	"github.com/abbyfluoroethane/bento/internal/types"
)

// AddSSHKey registers a public key for a user and returns its id.
func (s *Store) AddSSHKey(userID int64, publicKey, fingerprint, comment string) (int64, error) {
	res, err := s.db.Exec(`INSERT INTO ssh_keys (user_id, public_key, fingerprint, comment, created_at)
		VALUES (?, ?, ?, ?, ?)`,
		userID, publicKey, fingerprint, comment, fmtTime(s.now()))
	if err != nil {
		return 0, err
	}
	return res.LastInsertId()
}

// SSHKeyByFingerprint is the hot-path lookup the SSH frontend runs on
// every connection; idx_ssh_keys_fingerprint backs it (SPEC 12).
func (s *Store) SSHKeyByFingerprint(fingerprint string) (types.SSHKey, error) {
	row := s.db.QueryRow(`SELECT id, user_id, public_key, fingerprint, comment, created_at
		FROM ssh_keys WHERE fingerprint = ? ORDER BY id LIMIT 1`, fingerprint)
	return scanSSHKey(row)
}

// SSHKeysForUser lists the keys of one user.
func (s *Store) SSHKeysForUser(userID int64) ([]types.SSHKey, error) {
	rows, err := s.db.Query(`SELECT id, user_id, public_key, fingerprint, comment, created_at
		FROM ssh_keys WHERE user_id = ? ORDER BY id`, userID)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	var out []types.SSHKey
	for rows.Next() {
		key, err := scanSSHKey(rows)
		if err != nil {
			return nil, err
		}
		out = append(out, key)
	}
	return out, rows.Err()
}

// DeleteSSHKey removes one key of one user. The user scope stops a user
// from deleting another user's key by id.
func (s *Store) DeleteSSHKey(userID, keyID int64) error {
	res, err := s.db.Exec(`DELETE FROM ssh_keys WHERE id = ? AND user_id = ?`, keyID, userID)
	if err != nil {
		return err
	}
	n, err := res.RowsAffected()
	if err != nil {
		return err
	}
	if n == 0 {
		return ErrNotFound
	}
	return nil
}

func scanSSHKey(row rowScanner) (types.SSHKey, error) {
	var (
		key     types.SSHKey
		created string
	)
	err := row.Scan(&key.ID, &key.UserID, &key.PublicKey, &key.Fingerprint,
		&key.Comment, &created)
	if errors.Is(err, sql.ErrNoRows) {
		return types.SSHKey{}, ErrNotFound
	}
	if err != nil {
		return types.SSHKey{}, err
	}
	if key.CreatedAt, err = parseTime(created); err != nil {
		return types.SSHKey{}, err
	}
	return key, nil
}

File: crates/store/src/sshkeys.rs
use bento_types::SshKey;
use rusqlite::{OptionalExtension, params};

use crate::{Error, Result, Store, format_time, parse_time};

impl Store {
    /// Registers a public key for a user and returns its id.
    pub async fn add_ssh_key(
        &self,
        user_id: i64,
        public_key: impl Into<String>,
        fingerprint: impl Into<String>,
        comment: impl Into<String>,
    ) -> Result<i64> {
        let public_key = public_key.into();
        let fingerprint = fingerprint.into();
        let comment = comment.into();
        let now = self.clock();
        self.with_conn(move |conn| {
            conn.execute(
                "INSERT INTO ssh_keys \
                 (user_id, public_key, fingerprint, comment, created_at) VALUES (?, ?, ?, ?, ?)",
                params![
                    user_id,
                    public_key,
                    fingerprint,
                    comment,
                    format_time(now())?
                ],
            )?;
            Ok(conn.last_insert_rowid())
        })
        .await
    }

    /// Hot-path lookup run by the SSH frontend on every connection;
    /// `idx_ssh_keys_fingerprint` backs it (SPEC 12).
    pub async fn ssh_key_by_fingerprint(&self, fingerprint: impl Into<String>) -> Result<SshKey> {
        let fingerprint = fingerprint.into();
        self.with_conn(move |conn| {
            conn.query_row(
                "SELECT id, user_id, public_key, fingerprint, comment, created_at \
                 FROM ssh_keys WHERE fingerprint = ? ORDER BY id LIMIT 1",
                [fingerprint],
                scan_ssh_key,
            )
            .optional()?
            .ok_or(Error::NotFound)
        })
        .await
    }

    /// Lists one user's keys in insertion order.
    pub async fn ssh_keys_for_user(&self, user_id: i64) -> Result<Vec<SshKey>> {
        self.with_conn(move |conn| {
            let mut statement = conn.prepare(
                "SELECT id, user_id, public_key, fingerprint, comment, created_at \
                 FROM ssh_keys WHERE user_id = ? ORDER BY id",
            )?;
            let rows = statement.query_map([user_id], scan_ssh_key)?;
            Ok(rows.collect::<rusqlite::Result<Vec<_>>>()?)
        })
        .await
    }

    /// Removes one key belonging to one user. The user scope prevents a
    /// caller deleting another user's key by id.
    pub async fn delete_ssh_key(&self, user_id: i64, key_id: i64) -> Result<()> {
        self.with_conn(move |conn| {
            let changed = conn.execute(
                "DELETE FROM ssh_keys WHERE id = ? AND user_id = ?",
                params![key_id, user_id],
            )?;
            if changed == 0 {
                return Err(Error::NotFound);
            }
            Ok(())
        })
        .await
    }
}

fn scan_ssh_key(row: &rusqlite::Row<'_>) -> rusqlite::Result<SshKey> {
    let created: String = row.get(5)?;
    Ok(SshKey {
        id: row.get(0)?,
        user_id: row.get(1)?,
        public_key: row.get(2)?,
        fingerprint: row.get(3)?,
        comment: row.get(4)?,
        created_at: parse_time(5, &created)?,
    })
}

#[cfg(test)]
mod tests {
    use crate::Error;
    use crate::tests::{new_test_store, seed_store};

    #[tokio::test]
    async fn ssh_key_fingerprint_lookup() {
        let store = new_test_store().await;
        let (user, _) = seed_store(&store).await;
        let id = store
            .add_ssh_key(
                user.id,
                "ssh-ed25519 AAAA... alice@laptop",
                "SHA256:abcdef",
                "laptop",
            )
            .await
            .unwrap();
        store
            .add_ssh_key(
                user.id,
                "ssh-ed25519 BBBB... alice@desk",
                "SHA256:ghijkl",
                "desk",
            )
            .await
            .unwrap();
        let key = store.ssh_key_by_fingerprint("SHA256:abcdef").await.unwrap();
        assert_eq!(key.id, id);
        assert_eq!(key.user_id, user.id);
        assert_eq!(key.comment, "laptop");
        assert!(matches!(
            store.ssh_key_by_fingerprint("SHA256:missing").await,
            Err(Error::NotFound)
        ));
        assert_eq!(store.ssh_keys_for_user(user.id).await.unwrap().len(), 2);
        assert!(matches!(
            store.delete_ssh_key(user.id + 1, id).await,
            Err(Error::NotFound)
        ));
        store.delete_ssh_key(user.id, id).await.unwrap();
        assert!(matches!(
            store.ssh_key_by_fingerprint("SHA256:abcdef").await,
            Err(Error::NotFound)
        ));
    }
}
